From ee6bb43e8b35f2cc5df49dddf07f69e4b51e2e56 Mon Sep 17 00:00:00 2001
From: Eskil Abrahamsen Blomfeldt <eskil.abrahamsen-blomfeldt@qt.io>
Date: Fri, 25 Sep 2026 08:20:34 +0200
Subject: [PATCH] Revert "QTextEngine: keep the hb_buffer_t across
 shapeTextWithHarfbuzzNG calls"
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This reverts commit 8209078e0eb1f100f0f822d75856c9f557f60195. The commit
caused a large increase in memory usage of text layouts, and did not
include any benchmarks, bug report or other information that might be
used to justify such an increase.

Fixes: QTBUG-150286
Change-Id: I97aa8171f6737d21cb15883c5552ede2aaf79945
Reviewed-by: Eirik Aavitsland <eirik.aavitsland@qt.io>
Reviewed-by: Aurélien Brooke <aurelien@bahiasoft.fr>
(cherry picked from commit 56fae2893a89fe64f32f0b8897f434bc062f644a)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
Origin: upstream, https://invent.kde.org/qt/qt/qtbase/-/commit/ee6bb43e8b35f2cc5df49dddf07f69e4b51e2e56
Forwarded: not-needed
Last-Update: 2026-10-29
---
 src/gui/text/qtextengine.cpp | 24 ++++++++++--------------
 src/gui/text/qtextengine_p.h |  4 ----
 2 files changed, 10 insertions(+), 18 deletions(-)

diff --git a/src/gui/text/qtextengine.cpp b/src/gui/text/qtextengine.cpp
index 8dc820db27a..bac9791d9fc 100644
--- a/src/gui/text/qtextengine.cpp
+++ b/src/gui/text/qtextengine.cpp
@@ -1615,15 +1615,11 @@ int QTextEngine::shapeTextWithHarfbuzzNG(const QScriptItem &si, const ushort *st
 {
     uint glyphs_shaped = 0;
 
-    if (!buffer) {
-        buffer = hb_buffer_create();
-        hb_buffer_set_unicode_funcs(buffer, hb_qt_get_unicode_funcs());
-    }
-
+    hb_buffer_t *buffer = hb_buffer_create();
+    hb_buffer_set_unicode_funcs(buffer, hb_qt_get_unicode_funcs());
     hb_buffer_pre_allocate(buffer, itemLength);
     if (Q_UNLIKELY(!hb_buffer_allocation_successful(buffer))) {
         hb_buffer_destroy(buffer);
-        buffer = nullptr;
         return 0;
     }
 
@@ -1714,8 +1710,10 @@ int QTextEngine::shapeTextWithHarfbuzzNG(const QScriptItem &si, const ushort *st
                                           features.values().constData(),
                                           features.values().size(),
                                           shaper_list);
-            if (Q_UNLIKELY(!shapedOk))
+            if (Q_UNLIKELY(!shapedOk)) {
+                hb_buffer_destroy(buffer);
                 return 0;
+            }
 
             if (Q_UNLIKELY(HB_DIRECTION_IS_BACKWARD(props.direction)))
                 hb_buffer_reverse(buffer);
@@ -1728,8 +1726,10 @@ int QTextEngine::shapeTextWithHarfbuzzNG(const QScriptItem &si, const ushort *st
             num_glyphs = 1;
 
         // ensure we have enough space for shaped glyphs and metrics
-        if (Q_UNLIKELY(!ensureSpace(glyphs_shaped + num_glyphs)))
+        if (Q_UNLIKELY(!ensureSpace(glyphs_shaped + num_glyphs))) {
+            hb_buffer_destroy(buffer);
             return 0;
+        }
 
         // fetch the shaped glyphs and metrics
         QGlyphLayout g = availableGlyphs(&si).mid(glyphs_shaped, num_glyphs);
@@ -1789,6 +1789,8 @@ int QTextEngine::shapeTextWithHarfbuzzNG(const QScriptItem &si, const ushort *st
         glyphs_shaped += num_glyphs;
     }
 
+    hb_buffer_destroy(buffer);
+
     return glyphs_shaped;
 }
 
@@ -1832,12 +1834,6 @@ QTextEngine::~QTextEngine()
         delete layoutData;
     delete specialData;
     resetFontEngineCache();
-#if QT_CONFIG(harfbuzz)
-    if (buffer) {
-        hb_buffer_destroy(buffer);
-        buffer = nullptr;
-    }
-#endif
 }
 
 const QCharAttributes *QTextEngine::attributes() const
diff --git a/src/gui/text/qtextengine_p.h b/src/gui/text/qtextengine_p.h
index f27463f7728..31a12bd30e0 100644
--- a/src/gui/text/qtextengine_p.h
+++ b/src/gui/text/qtextengine_p.h
@@ -40,8 +40,6 @@
 #include <stdlib.h>
 #include <vector>
 
-struct hb_buffer_t;
-
 QT_BEGIN_NAMESPACE
 
 class QFontPrivate;
@@ -585,8 +583,6 @@ private:
     void indexFormats();
     void resolveFormats() const;
 
-    mutable hb_buffer_t *buffer = nullptr;
-
 public:
     bool atWordSeparator(int position) const;
 
